### PR TITLE
BUG: special.softmax: handle `np.inf`

### LINKED ISSUE #23225: BUG: softmax outputs nan for too large values
### Describe your issue.

`special.softmax` should be able to handle arbitrary values. Instead of evaluating to an array of `nan`s, it should find the largest value (if unique) and assign all mass to that one, similar as for
```ipython
In [1]: special.softmax([1., 1., 1e100])
Out[1]: array([0., 0., 1.])
```

### Reproducing Code Example

```python
import numpy as np
from scipy import special

special.softmax([1., 1., 1e1000000])
special.softmax([1., 1., np.inf])
```

### Error message

```shell
n/a
```

### SciPy/NumPy/Python version and system information

```shell
1.15.3 2.3.0 sys.version_info(major=3, minor=13, micro=5, releaselevel='final', serial=0)
{
  "Compilers": {
    "c": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "10.3.0",
      "commands": "cc"
    },
    "cython": {
      "name": "cython",
      "linker": "cython",
      "version": "3.0.12",
      "commands": "cython"
    },
    "c++": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "10.3.0",
      "commands": "c++"
    },
    "fortran": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "10.3.0",
      "commands": "gfortran"
    },
    "pythran": {
      "version": "0.17.0",
      "include directory": "C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pip-build-env-9_amgdyq\\overlay\\Lib\\site-packages/pythran"
    }
  },
  "Machine Information": {
    "host": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "windows"
    },
    "build": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "windows"
    },
    "cross-compiled": false
  },
  "Build Dependencies": {
    "blas": {
      "name": "scipy-openblas",
      "found": true,
      "version": "0.3.28",
      "detection method": "pkgconfig",
      "include directory": "C:/Users/runneradmin/AppData/Local/Temp/cibw-run-p9qsra0d/cp313-win_amd64/build/venv/Lib/site-packages/scipy_openblas32/include",
      "lib directory": "C:/Users/runneradmin/AppData/Local/Temp/cibw-run-p9qsra0d/cp313-win_amd64/build/venv/Lib/site-packages/scipy_openblas32/lib",
      "openblas configuration": "OpenBLAS 0.3.28 DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=24",
      "pc file directory": "D:/a/scipy/scipy"
    },
    "lapack": {
      "name": "scipy-openblas",
      "found": true,
      "version": "0.3.28",
      "detection method": "pkgconfig",
      "include directory": "C:/Users/runneradmin/AppData/Local/Temp/cibw-run-p9qsra0d/cp313-win_amd64/build/venv/Lib/site-packages/scipy_openblas32/include",
      "lib directory": "C:/Users/runneradmin/AppData/Local/Temp/cibw-run-p9qsra0d/cp313-win_amd64/build/venv/Lib/site-packages/scipy_openblas32/lib",
      "openblas configuration": "OpenBLAS 0.3.28 DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=24",
      "pc file directory": "D:/a/scipy/scipy"
    },
    "pybind11": {
      "name": "pybind11",
      "version": "2.13.6",
      "detection method": "config-tool",
      "include directory": "unknown"
    }
  },
  "Python Information": {
    "path": "C:\\Users\\runneradmin\\AppData\\Local\\Temp\\cibw-run-p9qsra0d\\cp313-win_amd64\\build\\venv\\Scripts\\python.exe",
    "version": "3.13"
  }
}
C:\Users\crunge\.virtualenvs\qit-GV4UT2CL\Lib\site-packages\scipy\__config__.py:154: UserWarning: Install `pyyaml` for better output
  warnings.warn("Install `pyyaml` for better output", stacklevel=1)
```

### PR DESCRIPTION
**Pull request by fbourgey**
Closes #23225

Enable `special.softmax` to handle inputs containing `np.inf`. 

- If exactly one entry `np.inf`, the result is 1 at the index and 0 elsewhere. 

- If multiple entries are `np.inf`, the probability mass is distributed evenly among them.
For example `softmax([1.0, np.inf, np.inf])` returns `softmax([0.0, 0.5, 0.5])` (see added tests)

### CODE REVIEW

File: scipy/special/tests/test_logsumexp.py [hunk 1]
```diff
@@ -478,3 +478,18 @@ def test_log_softmax_3d(self, xp):
...
+            ([1.0, np.inf, np.inf], [0.0, 0.5, 0.5]),
+            ([np.inf, np.inf, np.inf], [1 / 3, 1 / 3, 1 / 3]),
+            ([-np.inf, np.inf], [0.0, 1.0]),
+            ([np.inf], [1.0]),
+            ([-np.inf], [1.0]),
+        ],
+    )
+    def test_softmax_inf_inputs(self, xp, expected):
+        # Handle infinite inputs without producing NaNs - see gh-23225
+        xp_assert_close(softmax(np.asarray(xp)), np.asarray(expected), rtol=1e-13)
```
Comment (j-bowhay) [inline scipy/special/tests/test_logsumexp.py]: We need to use `xp` rather than `np` here otherwise the test won't work properly for other array types
```suggestion
        xp_assert_close(softmax(xp.asarray(xp)), xp.asarray(expected), rtol=1e-13)
```

File: scipy/special/_logsumexp.py [hunk 2]
```diff
@@ -342,8 +342,10 @@ def softmax(x, axis=None):
     """
     xp = array_namespace(x)
     x = xp.asarray(x)
+    x = xp.where(xp.isinf(x), 1e308 * xp.sign(x), x)
```
Comment (j-bowhay) [inline scipy/special/_logsumexp.py]: Since the `dtype` is not necessarily `float64` we should use https://data-apis.org/array-api/latest/API_specification/generated/array_api.finfo.html#array_api.finfo
to get the correct max

File: scipy/special/tests/test_logsumexp.py [hunk 3]
```diff
@@ -395,6 +394,24 @@ def test_softmax_array_like(self):
...
+            ([-np.inf], [1.0]),
+        ],
+    )
+    @pytest.mark.parametrize("dtype", ["float32", "float64"])
+    def test_softmax_inf_inputs(self, xp_input, expected, dtype, xp):
+        dtype = getattr(xp, dtype)
+        # Handle infinite inputs without producing NaNs - see gh-23225
+        xp_assert_close(softmax(xp.asarray(xp_input, dtype=dtype)),
+                        xp.asarray(expected, dtype=dtype), rtol=1e-13)
+
```
Comment (tylerjereddy) [inline scipy/special/tests/test_logsumexp.py:518]: Just checking, but have we discussed if it makes sense to deviate from `pytorch` `softmax`? I don't think I saw that in skimming over the cognate issue.

In checking locally, with a temporary test shim as below the fold,

<details>

```diff
--- a/scipy/special/tests/test_logsumexp.py
+++ b/scipy/special/tests/test_logsumexp.py
@@ -5,7 +5,7 @@ import pytest
 import numpy as np
 
 from scipy._lib._array_api import (is_array_api_strict, make_xp_test_case,
-                                   xp_default_dtype, xp_device)
+                                   xp_default_dtype, xp_device, is_torch)
 from scipy._lib._array_api_no_0d import (xp_assert_equal, xp_assert_close,
                                          xp_assert_less)
 
@@ -409,8 +409,10 @@ class TestSoftmax:
     def test_softmax_inf_inputs(self, xp_input, expected, dtype, xp):
         dtype = getattr(xp, dtype)
         # Handle infinite inputs without producing NaNs - see gh-23225
-        xp_assert_close(softmax(xp.asarray(xp_input, dtype=dtype)),
-                        xp.asarray(expected, dtype=dtype), rtol=1e-13)
+        actual = softmax(xp.asarray(xp_input, dtype=dtype))
+        if is_torch(xp):
+            expected = xp.softmax(xp.asarray(xp_input, dtype=dtype), dim=0, dtype=dtype)
+            xp_assert_close(actual, expected, rtol=1e-13)
```

</details>

`spin test -t scipy/special/tests/test_logsumexp.py -b torch -- -k "test_softmax_inf_inputs"` fails against the torch `softmax` proper.

But it passes (matches `torch`) on current SciPy `main` (if I back out the current proposed patch).

I'm not saying `torch` is the "oracle of truth" here, but `softmax` is of course heavily used in ML, so one might think they've thought about this?

[The test changes I made above are not desirable, they were just a quick way to mutate this PR branch to temporarily compare against `torch`]

Comment (steppi) [reply scipy/special/tests/test_logsumexp.py:518]: > Just checking, but have we discussed if it makes sense to deviate from `pytorch` `softmax`? I don't think I saw that in skimming over the cognate issue.

I don't have a strong opinion here. In principle it seems like a good idea to reduce the number of moving parts in these tests, but I think this particular case does fall under the following carve-out I want to add to the docs in #24130:

https://github.com/scipy/scipy/blob/194ab2fde226d25543af35249379bc9d17268c86/doc/source/dev/api-dev/array_api.rst?plain=1#L1028-L1031

**Comment by tylerjereddy** [comment]
I just meant to bring up the discussion point that `pytorch` already has the function `softmax` and it behaves differently than what this PR proposes to do with our own version of `softmax`, which is a point arguably-independent from the fact that we're supporting `torch` objects in our own `softmax`.

I wasn't proposing that we "dispatch" to the `torch` `softmax` (though I believe we do/did have some places where we *do* dispatch to backends for some functionality), just that we have the brief discussion on whether we have a good reason to deviate from their approach here.

**Comment by steppi** [comment]
> I just meant to bring up the discussion point that `pytorch` already has the function `softmax` and it behaves differently than what this PR proposes to do with our own version of `softmax`, which is a point arguably-independent from the fact that we're supporting `torch` objects in our own `softmax`.
> 
> I wasn't proposing that we "dispatch" to the `torch` `softmax` (though I believe we do/did have some places where we _do_ dispatch to backends for some functionality), just that we have the brief discussion on whether we have a good reason to deviate from their approach here.

Sorry, I see what you're saying now. I think SciPy does actually delegate to ``torch.special.softmax`` on PyTorch arrays. I'll look into comparing the two approaches, and can make a ``torch`` issue if I think the approach proposed here is better.

File: scipy/special/_logsumexp.py [hunk 4]
```diff
@@ -341,9 +341,11 @@ def softmax(x, axis=None):
 
     """
     xp = array_namespace(x)
-    x = xp.asarray(x)
+    x = xp_promote(x, force_floating=True, xp=xp)
+    x = xp.where(xp.isinf(x), xp.finfo(x.dtype).max * xp.sign(x), x)
```
Comment (WarrenWeckesser) [inline scipy/special/_logsumexp.py]: This will give an incorrect result if `x` already contains `xp.finfo(x.dtype).max`.  E.g. consider `softmax([np.finfo(np.float64).max, np.inf])`.  The expected answer is `[0.0, 1.0]`, not `[0.5, 0.5]`.

**Comment by WarrenWeckesser** [comment]
For a change like this, a brief description of how you are resolving the issue would be appreciated, either in the opening comment of this PR, or even better in the commit message that makes the change.  Of course, it can be inferred from the code, but including an explicit statement of the API design decision is better.  It allows reviewers to understand the *intent* of the change, and that makes it easier to assess correctness.

Handling a single positive `inf` so that the output is 1 at the location of the `inf` and the other values are 0 is a natural choice.

There was no resolution in gh-23225 of how multiple positive `inf`s should be handled. With this PR, the output is evenly distributed over the locations of the `inf`s.  E.g. (from one of the tests), `softmax([1.0, np.inf, np.inf])` returns `[0.0, 0.5, 0.5]`.   But the "correct" result in this case is ambiguous, and an argument can be made that the result should be `nan` (as it currently is before this PR) [1].   Is there prior art for the "evenly distributed" behavior?  @tylerjereddy mentioned PyTorch; how does the PyTorch implementation handle one or more `inf`s in the input?

I lean slightly towards `nan` in the ambiguous case, but if y'all prefer the "evenly distributed" behavior, I wouldn't fight it.  Whatever behavior is chosen, it should be documented in the docstring, perhaps with just a couple sentences in the Notes section.

-----

[1] Zen of Python: *In the face of ambiguity, refuse the temptation to guess.* In my experience, that has turned out to be one of the actually useful "koans" in the Zen of Python.

**Comment by fbourgey** [comment]
@WarrenWeckesser thanks, I've updated my opening PR comment.

Regarding torch, this is what I have (torch version = '2.5.1')

```python
x = torch.tensor([1.0, 1.0, torch.inf], dtype=torch.float64)
print(F.softmax(x, dim=0))
tensor([nan, nan, nan], dtype=torch.float64)

x = torch.tensor([1.0, torch.inf, torch.inf], dtype=torch.float64)
print(F.softmax(x, dim=0))
tensor([nan, nan, nan], dtype=torch.float64)

x = torch.tensor([1.0, 1.0, 1e100], dtype=torch.float64)
print(F.softmax(x, dim=0))
tensor([0., 0., 1.], dtype=torch.float64)

x = torch.tensor([1.0, 1.0, 1e100], dtype=torch.float32)
print(F.softmax(x, dim=0))
tensor([nan, nan, nan])

x = torch.tensor([1.0, torch.finfo().max, torch.finfo().max], dtype=torch.float64)
print(F.softmax(x, dim=0))
tensor([0.0000, 0.5000, 0.5000], dtype=torch.float64)
```

**Comment by fancidev** [comment]
To me there is a mathematical aspect and a numerical aspect to this question.

The mathematical aspect is how to define the softmax function when one or more of the arguments is infinity. If the softmax is a function $\mathbb{R}^n \mapsto [0,1]^n$, then $\pm \infty$ are out of the domain and hence undefined. Therefore, if we want to return something other than nan, we have to invent a definition. The proposed PR could be rationalized by fixing the finite arguments and let the non-finite element(s) tend to $+\infty$ *at the same speed*. When there are more than one infinities this treatment seems somewhat unusual (the limit should not depend on how it's approached), though there have been exceptions (e.g. the atan2 function of IEEE 754 returns 0 at 0, 0 though it is mathematically undefined).

The numerical aspect is related to how the infinity is obtained. It could either mean the mathematical $\infty$, or it could represent an intermediate result that is larger than the largest representable finite floating point value (fmax). For the latter case, if there is a single inf, it may be verified that the proposed PR works correctly. If there are two or more infinities, however, we could only conclude that they all exceed the representable range, but no idea which one is actually larger. Evenly distributing the weights seems like a best-effort solution, but it doesn't seem entirely obvious to me either.

**Comment by j-bowhay** [comment]
@Uroc327 what was your use case for needing infinite inputs?

**Comment by fbourgey** [comment]
How should we proceed from here?

**Comment by j-bowhay** [comment]
I think we still need to make a decision about the desired behaviour. @Uroc327 do you have anymore information about your use case?

**Comment by Uroc327** [comment]
Personally, I don't think, the general behaviour should depend on a specific use case here. I think, if an answer depends on the use, e.g., on how quickly arguments grow to infinity relative to each other, the result should indeed be something like nan and such a case should be handled by the caller. 

My use case was working with probability values (likelihoods, log likelihoods, log likelihood ratios) and information theoretic quantities (entropies and similar), where the soft max can come up in different formulae.

**Comment by fbourgey** [comment]
Any further thoughts on this? I’m happy to close the PR if we can’t reach a consensus, especially if we prefer to stay aligned with PyTorch.

**Comment by WarrenWeckesser** [comment]
I'll convert my previous comments into a concrete proposal for handling `inf`:

* When the input contains a *single* positive `inf`, the output is 1 there and 0 elsewhere.
* When there is more than one positive `inf`, the output is all `nan`, and the function call generates a warning.
* If *all* the values are `-inf`, the output is all `nan`, and the function call generates a warning. *(Added this after the initial comment.)*

If there are no counter-proposals in the next 48 hours, we can accept that behavior and work on implementing it. (That's a relatively short time, but the PR has been open since December, and I suspect anyone who is interested has already chimed in.)

**Comment by WarrenWeckesser** [comment]
I updated my comment to include the case where *all* the inputs are `-inf`.  This should be handled the same as the case where more than one input is `+inf`: generate a warning and return all `nan`.

**Comment by fbourgey** [comment]
Thanks @WarrenWeckesser, I am going to work on this in the coming days. It might be worth mentioning it in the [softmax documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.softmax.html) as well.

**Comment by fbourgey** [comment]
@WarrenWeckesser I have tried to implement your suggestions.

Should we test against the warnings? If so, what is the best way to achieve this with the `array_api`? Maybe `eager_warns` from `scipy._lib._array_api`?

**Comment by fbourgey** [comment]
@WarrenWeckesser can you please take a look when you get a chance?

**Comment by fbourgey** [comment]
@WarrenWeckesser @steppi @tylerjereddy @j-bowhay. Do we still think this should be added?
